src/Shared/PackageManagers/BaseProjectManager.cs

```diff
@@ -291,6 +291,17 @@ public virtual Task<IEnumerable<string>> EnumerateVersionsAsync(PackageURL purl,
         {
             throw new NotImplementedException("BaseProjectManager does not implement EnumerateVersions.");
         }
+        
+        /// <summary>
+        /// Gets the <see cref="DateTime"/> a package version was published at.
+        /// </summary>
+        /// <param name="purl">Package URL specifying the package. Version is mandatory.</param>
+        /// <param name="useCache">If the cache should be used when looking for the published time.</param>
+        /// <returns>The <see cref="DateTime"/> when this version was published, or null if not found.</returns>
+        public virtual Task<DateTime?> GetPublishedAtAsync(PackageURL purl, bool useCache = true)
+        {
+            throw new NotImplementedException("BaseProjectManager does not implement GetPublishedAtAsync.");
+        }
```

Comment:
Nit: I wonder whether it would make sense to hold off on adding new methods like this to `BaseProjectManager` until there is 'critical mass' per se of implementations for them for the specific package managers.

Advertising a method like this for all package managers when it only actually works for a small subset is not a great developer experience to have by default.

Reply:
I get what you're saying. As far as I see it then we have two options. We have this one. And the other one is the consumer has to cast the `BaseProjectManager` to one of the 3 that implement the method then they can call it. Personally I would rather do it this way, but what do you think?

Reply:
CFS is the only consumer of this method today right? And if I recall correctly, we are planning to instantiate the correct type of project manager directly anyway?

If that's true, then `BaseProjectManager` only really serves callers who want to work with more than one type of package at a time / who do not know which type of project manager instance they are getting. If we don't really have any consumers of this data yet from that perspective, I wonder if that gives us time to hold off on adding it to `BaseProjectManager` until that use case either materializes, or when we finish implementing this for many/most package managers.

Reply:
Yeah that's a good point. I will have it in just the ones we need it in.

Reply:
Is the published time available in the metadata from the metadata method? The default behavior could be to call the metadata method and then return the time.

Reply:
Ah I see metadata method isn't standard in the format it returns. That seems to be an opportunity for a metadata class with basic things like the requested purl, publish time etc. which would enable a default implementation of the publish time by calling the base metadata call - overrideable if there is a more efficient way for that specific manager.

Reply:
Realistically we could just do something like this on the `BaseProjectManager` in this method. And then it relies on `GetPackageMetadataAsync` being implemented but otherwise works as expected:
``` cs
PackageMetadata? metadata = await this.GetPackageMetadataAsync(purl, useCache);
string? uploadTime = metadata?.UploadTime;
if (uploadTime == null)
{
  return null;
}
return DateTime.Parse(uploadTime);

Reply:
For how many package managers is the `metadata?.UploadTime` value non-null? If many/most provide a value, then go for it. If only a handful do, then I'm not sure this would be all that better than throwing `NotImplementedException` as of now.

Reply:
Only the 3 do right now. I'll just do this, but on the Manager implementation side vs doing all the parsing. It simplifies it a lot.

Reply:
(Also, double checking an assumption - `DateTime.Parse` will be able to understand the string value in `metadata?.UploadTime` for all package managers right?)

Reply:
Yes, and I confirmed it in the unit tests.

Reply:
It seems to me the upload time value in the metadata object should already be a DateTime.

Reply:
I could change it to parse it elsewhere that makes sense. I'll go do it real quick.

Reply:
What about having this in the base, once the metadata is returned as an object?
```c#
public async DateTime GetPublishTimeAsync(PackagURL Purl) => await GetMetaDataAsync(Purl)?.PublishTime
```

This requires having a metadata object be returned
```c#
public class Metadata
{
    public DateTime PublishTime {get; set;}
    public string Raw { get; set }
}
```

Reply:
Yeah that can be something we look into when we do the larger rework to Metadata.